sequencer/src/header.rs

```diff
@@ -244,21 +260,47 @@ impl BlockHeader<SeqTypes> for Header {
         payload_commitment: VidCommitment,
         metadata: <<SeqTypes as NodeType>::BlockPayload as BlockPayload>::Metadata,
     ) -> Self {
-        // The HotShot APIs should be redesigned so that
-        // * they are async
-        // * new blocks being created have access to the application state, which in our case could
-        //   contain an already connected L1 client.
-        // For now, as a workaround, we will create a new L1 client based on environment variables
-        // and use `block_on` to query it.
-        let l1 = instance_state.l1_client().snapshot().await;
+        // Fetch the latest L1 snapshot.
+        let l1_snapshot = instance_state.l1_client().snapshot().await;
+        // Fetch the new L1 deposits between parent and current finalized L1 block.
+        let l1_deposits = fetch_fee_receipts(
+            parent_leaf.get_block_header().l1_finalized,
+            l1_snapshot.finalized,
+        );
+
+        // Find missing fee state entries
+        let missing_accounts = parent_state.forgotten_accounts(
+            std::iter::once(FeeAccount::from(instance_state.builder_address.address()))
+                .chain(l1_deposits.iter().map(|info| info.account())),
+        );
+
+        // Fetch missing fee state entries
+        let missing_account_proofs = instance_state
+            .peers
+            .fetch_accounts(
+                parent_leaf.get_view_number(),
+                parent_state.fee_merkle_tree.commitment(),
+                missing_accounts,
+            )
+            .await;
+
+        // Insert missing fee state entries
+        let mut state = parent_state.clone();
```

Comment:
Since we're cloning `state` here we could pass it into `from_info` by value and avoid a second clone (although `clone` should be fairly cheap now that stuff is `Arc`ed)